File: algorithms/[888]公平的糖果交换/solution.py
from typing import List, Dict


class Solution:
    def fairCandySwap(self, A: List[int], B: List[int]) -> List[int]:
        t1 = sum(A)
        t2 = sum(B)
        avg = (t1 + t2) // 2

        d: Dict[int, int] = {}
        for ele in A:
            another = ele + avg - t1
            d.setdefault(another, ele)

        for ele in B:
            if ele in d:
                return [d[ele], ele]

File: algorithms/[888]公平的糖果交换/solution.ts
// topics = ["哈希表"]

/**
 * Hash Table
 * time O(n + m), space O(n), n 和 m 分别为 A 和 B 的长度
 */
function fairCandySwap(A: number[], B: number[]): number[] {
  const t1 = A.reduce((a, b) => a + b, 0);
  const t2 = B.reduce((a, b) => a + b, 0);
  const avg = Math.floor((t1 + t2) / 2);

  // 不考虑索引的情况下优先使用 Set
  const setA = new Set(A);
  for (const val of B) {
    const another = val + avg - t2;
    if (setA.has(another)) {
      return [another, val];
    }
  }

  return [];
}
